File: 2015/06.rb
fn = open("2015/06.input")
# 0, off, 1, on, 2, toggle
flags = ['f', 'n', ' ']
input = fn.each_line.map(&:chomp).map{|l|
  [flags.index(l[6])] + l.scan(/\d+/).map(&:to_i)
}


# input = [
#   [1, 0, 0, 9, 9],
#   [2, 0, 0, 9, 0],
#   [0, 0, 9, 9, 9]
# ]
N = 1000
lights = Array.new(N*N, 0)

input.each{|l|
  op, l, t, r, b = l
  (t..b).each{|y| 
    (l..r).each{|x|
      p = y * N + x
      lights[p] = 0 if op == 0
      lights[p] = 1 if op == 1
      lights[p] = 1 - lights[p] if op == 2
    }
  }
}

p lights.count(1)
# p lights

File: 2015/06.js
import { range } from 'lodash'

const flags = ['f', 'n', ' ']
const processData = s => [
  flags.indexOf(s[6]),
  ...s.scan(/(\d+)/g).map(Number)
]

const N = 1000
const initLights = () => Array.new(N * N, 0)

const toggle = (v, op) => ((op > 1) ? 1 - v : op)
const bright = (v, op) => ((op === 0) ? Math.max(0, v - 1) : v + op)
const toggleLights = toggleFn => (lights, v) => {
  const [op, l, t, r, b] = v
  range(t, b + 1).forEach(y => {
    const row = y * N
    range(l, r + 1).forEach(x => {
      const p = row + x
      lights[p] = toggleFn(lights[p], op)
    })
  })
  return lights
}

const prepare = data => data.map(processData)
const part1 = data => data
  .reduce(toggleLights(toggle), initLights())
  .filter(Math.sign).length
const part2 = data => data
  .reduce(toggleLights(bright), initLights())
  .sum()

export default () => ({ prepare, part1, part2 })
